### PR TITLE
[Android] Move nameindex implementation

### LINKED ISSUE #29483: Investigate whether the cross-platform code can just work without needing the Android ifdefs when encoutering if_nameindex and if_freenameindex  in PR29244
>  It is not easy to change this code to be implemented in the Android folder.

I wasn't asking for that.  I was asking for a separate file that is compiled only on Android.  It can live in this directory.

I didn't realize we had used to have this workaround in this form, but the question remains: if the issue is that `if_nameindex` and `if_freenameindex` don't exist, then why are we not naming our implementations _that_, and then the cross-platform code can just work without needing the Android ifdefs.

_Originally posted by @bzbarsky-apple in https://github.com/project-chip/connectedhomeip/pull/29244#discussion_r1339576158_

### PR DESCRIPTION
**Pull request by joonhaengHeo**
Fixes #29483 

1. Remove android ifdef in InetInterface.cpp
2. Move implementation to platform/android
3. Add feature in inet/BUILD.gn

### CODE REVIEW

**Review by yufengwangca — APPROVED** [review]

**Comment by yufengwangca** [comment]
Typically, we don't append the "Default" postfix to the command implementation in the src/inet/InetInterfaceImplDefault.cpp file. Instead, we implement files with the same names under src/platform/<xxx>/ to override the default implementation.

File: src/platform/android/InetInterfaceImpl.cpp [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+#include <ifaddrs.h>
+#include <lib/support/CHIPMem.h>
+#include <lib/support/CodeUtils.h>
+#include <net/if.h>
+namespace chip {
+namespace Inet {
+
+using namespace chip;
+
+void if_freenameindexImpl(struct if_nameindex * inArray)
```
Comment (bzbarsky-apple) [inline src/platform/android/InetInterfaceImpl.cpp:30]: Can we just name this `if_freenameindex`, put it in the global namespace, and not need the "Impl" shimming?  Similar for `if_nameindex`.

That is, is Android just missing these pieces, in which case we should just implement them, or are they there but wrong?

Comment (joonhaengHeo) [reply src/platform/android/InetInterfaceImpl.cpp:30]: The function exists, but it did not work, so it was reimplemented. Therefore, implementation with the same function name is not possible. Could you recommend an suitable function name?

File: src/inet/InetInterfaceImpl.h [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+#pragma once
+
+#include <system/SystemConfig.h>
+
+#if CHIP_SYSTEM_CONFIG_USE_SOCKETS && CHIP_SYSTEM_CONFIG_USE_BSD_IFADDRS
+#include <net/if.h>
+
+namespace chip {
+namespace Inet {
+// Define if_nameindex to be implemented in the platform layer
```
Comment (bzbarsky-apple) [inline src/inet/InetInterfaceImpl.h:27]:
```suggestion
// Some platforms have broken if_nameindex and if_freenameindex implementations.
// Allow platforms to either wrap a non-broken platform implementation or have a custom
// implementation.
```

File: src/platform/android/InetInterfaceImpl.cpp (hunk 1 above)
Comment (bzbarsky-apple) [reply src/platform/android/InetInterfaceImpl.cpp:30]: If they exist but are broken, then these names seem fine.  Thank you for the clarification!

**Review by bzbarsky-apple — APPROVED** [review]

**Pull request merged**